rpcserver.go

```diff
@@ -840,6 +844,81 @@ func (r *rpcServer) ListUnspent(ctx context.Context,
 	return resp, nil
 }
 
+// PendingSweeps returns lists of on-chain outputs that lnd is currently
+// attempting to sweep within its central batching engine. Outputs with similar
+// fee rates are batched together in order to sweep them within a single
+// transaction. The fee rate of each sweeping transaction is determined by
+// taking the average fee rate of all the outputs it's trying to sweep.
+func (r *rpcServer) PendingSweeps(ctx context.Context,
+	in *lnrpc.PendingSweepsRequest) (*lnrpc.PendingSweepsResponse, error) {
+
+	// Retrieve all of the outputs the UtxoSweeper is currently trying to
+	// sweep.
+	pendingInputs, err := r.server.sweeper.PendingInputs()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert them into their respective lnrpc format.
+	rpcPendingSweeps := make([]*lnrpc.PendingSweep, 0, len(pendingInputs))
+	for _, pendingInput := range pendingInputs {
+		var witnessType lnrpc.WitnessType
+		switch pendingInput.WitnessType() {
+		case input.CommitmentTimeLock:
+			witnessType = lnrpc.WitnessType_COMMITMENT_TIME_LOCK
+		case input.CommitmentNoDelay:
+			witnessType = lnrpc.WitnessType_COMMITMENT_NO_DELAY
+		case input.CommitmentRevoke:
+			witnessType = lnrpc.WitnessType_COMMITMENT_REVOKE
+		case input.HtlcOfferedRevoke:
+			witnessType = lnrpc.WitnessType_HTLC_OFFERED_REVOKE
+		case input.HtlcAcceptedRevoke:
+			witnessType = lnrpc.WitnessType_HTLC_ACCEPTED_REVOKE
+		case input.HtlcOfferedTimeoutSecondLevel:
+			witnessType = lnrpc.WitnessType_HTLC_OFFERED_TIMEOUT_SECOND_LEVEL
+		case input.HtlcAcceptedSuccessSecondLevel:
+			witnessType = lnrpc.WitnessType_HTLC_ACCEPTED_SUCCESS_SECOND_LEVEL
+		case input.HtlcOfferedRemoteTimeout:
+			witnessType = lnrpc.WitnessType_HTLC_OFFERED_REMOTE_TIMEOUT
+		case input.HtlcAcceptedRemoteSuccess:
+			witnessType = lnrpc.WitnessType_HTLC_ACCEPTED_REMOTE_SUCCESS
+		case input.HtlcSecondLevelRevoke:
+			witnessType = lnrpc.WitnessType_HTLC_SECOND_LEVEL_REVOKE
+		case input.WitnessKeyHash:
+			witnessType = lnrpc.WitnessType_WITNESS_KEY_HASH
+		case input.NestedWitnessKeyHash:
+			witnessType = lnrpc.WitnessType_NESTED_WITNESS_KEY_HASH
+		default:
+			rpcsLog.Warnf("Unhandled witness type %v",
+				pendingInput.WitnessType())
+		}
+
+		opHashStr := pendingInput.OutPoint().Hash.String()
+		opHashBytes, _ := hex.DecodeString(opHashStr)
+		amountSat := uint32(pendingInput.SignDesc().Output.Value)
+		satPerByte := uint32(pendingInput.FeeRate.FeePerKVByte() / 1000)
+		broadcastAttempts := uint32(pendingInput.PublishAttempts)
+		nextBroadcastHeight := uint32(pendingInput.MinPublishHeight)
+
+		rpcPendingSweeps = append(rpcPendingSweeps, &lnrpc.PendingSweep{
+			Outpoint: &lnrpc.OutPoint{
+				TxidStr:     opHashStr,
+				TxidBytes:   opHashBytes,
```

Comment:
We still need to populate both from the server's perspective as we don't know ahead of time which one the client is interested in.

Reply:
I mean just don't offer the string based one. This is a machine to machine interface. But can also understand that this may not be this pr's problem to address.

Reply:
Ah yes, I'm not a fan of the duplicate fields either. If we're worried about showing raw bytes as base64 in `lncli`, we can just map that to hex in the response.

Reply:
Atm we only use the str variants if we're accepting them within the path URL for the REST interface.

Reply:
So since this is just an output, I think we only need to set one of them.

Reply:
Removed the string one.